enhancements/microshift/microshift-greenboot.md

```diff
@@ -0,0 +1,350 @@
+by a Device Administrator, End User, etc.,
+the [greenboot-healthcheck.service](https://github.com/fedora-iot/greenboot#how-does-it-work)
+runs to verify the system validity.
+
+If the system validity procedures are successful, the version of the software
+is declared valid for end users. Otherwise, `greenboot` enters a system
+verification loop by repeating reboots up-to `GREENBOOT_MAX_BOOT_ATTEMPTS` times
+before attempting to rollback to the previous version of the software.
+
+```mermaid
+graph TD
+    A[Software vN] --> |Software Upgrade| B[Reboot Pending]
+    B --> |System Reboot| D{Validate Software vN+1}
+    D --> |Valid| E[Software vN+1]
+    D --> |Invalid| F[Verification Loop]
+    F --> |Up to Max Reboots|D
+    F --> |All Failed Attempts| G[Rollback]
+    G --> A
+```
+
+### API Extensions
+
+None
+
+### Implementation Details
+
+The MicroShift core RPM package will include the `microshift-greenboot-required.sh`
+health check script to validate that all the required MicroShift services are
+up and running. There will be no explicit dependency on the `greenboot` RPM.
+The script will be installed into the `/etc/greenboot/check/required.d` directory
+and it will not be executed during the system boot in case the `greenboot` package
+is not present.
+
+> If other scripts are present in the `/etc/greenboot/check/required.d` directory,
+> they will be executed by `greenboot` in the alphanumeric order.
+
+The script performs the following sanity checks:
+* Exit if the current user is not `root`
+* Exit if MicroShift service is not enabled
+* Wait for the MicroShift service to become active
+
+If the `greenboot-default-health-check` RPM subpackage is installed, it already
+includes [health check scripts](https://github.com/fedora-iot/greenboot#health-checks-included-with-subpackage-greenboot-default-health-checks)
+verifying that DNS and `ostree` services can be accessed.
+
+The health check script will implement validations for the MicroShift health
+endpoints and its services running as pods in the following namespaces:
+- openshift-ovn-kubernetes
+- openshift-service-ca
+- openshift-ingress
+- openshift-dns
+- openshift-storage
+
+> When pluggable CNI and CSI functionality is implemented, the `openshift-ovn-kubernetes`
+> and `openshift-storage` namespace name selection need to be conditional on
+> the loaded CNI and CSI plugins. The plugin developers may choose to add plugin-
+> specific health check scripts.
+
+Exiting the script with a non-zero status will have the boot declared as failed.
+Greenboot redirects all the script output to the system log, accessible via the
+`journalctl -u greenboot-healthcheck.service` command.
+
+|Validation                                           |Pass  |Fail  |
+|-----------------------------------------------------|------|------|
+|Check microshift.service is enabled                  |Next  |exit 0|
+|Wait for microshift.service to be active (!failed)   |Next  |exit 1|
```

Comment:
Instead of 'active' you probably want to test for it reporting 'ready', which covers some of the following points already.

Reply:
"active" is the status of the service. The service becomes active only when all the MicroShift components are initialized. That's "is-ready" implicitly.

Also, there seems to be no explicit "is-ready" state for systemd, only is-active, is-enabled and is-failed.

Reply:
Yes, you're right. I wrongly assumed "ready" would be a different state.